complex/ft_msg.c

```diff
@@ -217,13 +233,17 @@ int ft_send_msg(void)
 {
 	int ret;
 
-	while (!ft_tx.credits) {
-		ret = ft_comp_tx(FT_COMP_TO);
-		if (ret)
-			return ret;
+	if (test_info.class_function != FT_FUNC_INJECT &&
+		test_info.class_function != FT_FUNC_INJECTDATA) {
+		while (!ft_tx.credits) {
+			ret = ft_comp_tx(FT_COMP_TO);
+			if (ret)
+				return ret;
+		}
+
+		ft_tx.credits--;
 	}
 
-	ft_tx.credits--;
```

Comment:
The changes to ft_send_msg can be avoided by incrementing tx credits after calling inject.  We may also want to consider calling fi_tx_size_left() to determine if an inject call can be inserted successfully. Maybe something like:

do {
   ret = fi_inject()
   if (ret == 0 || ret != FI_EAGAIN)
      break;
   ft_comp_tx(...)   /* for manual progress mode */
   ret = fi_tx_size_left()
} while (ret == 0);